swap/src/monero/wallet.rs

```diff
@@ -315,6 +273,61 @@ impl Wallet {
     }
 }
 
+/// Wait until the specified transfer has been completed or failed.
+pub async fn watch_for_transfer(
+    wallet: Arc<Mutex<Wallet>>,
+    request: WatchRequest,
+) -> Result<(), InsufficientFunds> {
+    watch_for_transfer_with(wallet, request, None).await
+}
+
+/// Wait until the specified transfer has been completed or failed and listen to each new confirmation.
+#[allow(clippy::too_many_arguments)]
+pub async fn watch_for_transfer_with(
+    wallet: Arc<Mutex<Wallet>>,
+    request: WatchRequest,
+    listener: Option<ConfirmationListener>,
+) -> Result<(), InsufficientFunds> {
+    let WatchRequest {
+        conf_target,
+        public_view_key,
+        public_spend_key,
+        transfer_proof,
+        expected,
+    } = request;
+
+    let txid = transfer_proof.tx_hash();
+
+    tracing::info!(
+        %txid,
+        target_confirmations = %conf_target,
+        "Waiting for Monero transaction finality"
+    );
+
+    let address = Address::standard(
+        wallet.lock().await.network,
+        public_spend_key,
+        public_view_key.into(),
+    );
+
+    let check_interval = tokio::time::interval(wallet.lock().await.sync_interval.div(10));
```

Comment:
e.g having to acquire a lock at this step is a bit overkill as we are simply reading an attribute.